File: Solutions/Source2/Arrays/SecondLargest.py
'''
By: Marcos Gil
Solution for this practice problem
https://practice.geeksforgeeks.org/problems/second-largest/0
'''

'''
Name: getList
Purpose: Get values for list from the user
 In: N/A
 In/Out: N/A
 Out: list = list of values as integers
'''
def getList():
    userValues = input("Enter space seperated numbers as list values: ").split(" ")
    list = [int(x) for x in userValues]
    return list

'''
Name: secondLargest
Purpose: Find second largest value in list
 In: list = list of values
 In/Out: N/A
 Out: int = second largest value in list
'''
def secondLargest(list):
    largest = -9999999999999
    secondLargest = -9999999999999

    for i in range(0, len(list)):
        if list[i] > largest:
            secondLargest = largest
            largest = list[i]
        elif list[i] > secondLargest and list[i] != largest:
            secondLargest = list[i]

    return secondLargest

def main():
    list = getList()
    result = secondLargest(list)
    if (len(list) < 2):
        print("Not enough values entered.")
    else:
        print("The second largest value in", list, "is", result)

main()

File: Solutions/Source2/Arrays/SecondLargest.java
/*
 * By: Marcos Gil
 * Solution for this practice problem
 * https://practice.geeksforgeeks.org/problems/second-largest/0
 */
import java.util.Scanner;
import java.util.Arrays;

public class SecondLargest{

  /*
  Name: getArray
  Purpose: Get the length and values of an array from the user
    In: N/A
    In/Out: N/A
    Out: int[] array
  */
  public static int[] getArray(){

    int lengthOfArray = -1;
    Scanner userInput = new Scanner(System.in);
    System.out.print("Please enter the length of the array: ");
    lengthOfArray = userInput.nextInt();

    int[] array = new int[lengthOfArray];

    for (int i = 0; i < lengthOfArray; i++){
      System.out.print("Enter value " + (i + 1) + " of the array: ");
      array[i] = userInput.nextInt();
    }

    return array;
  }

  /*
  Name: secondLargest
  Purpose: Find second largest value in the array
    In: int[] arr
    In/Out: N/A
    Out: int secondLargest
  */
  public static int secondLargest(int[] arr){

    int largest = Integer.MIN_VALUE;
    int secondLargest = Integer.MIN_VALUE;

    for (int i = 0; i < arr.length; i++){
      if (arr[i] > largest){
        secondLargest = largest;
        largest = arr[i];
      } else if (arr[i] > secondLargest && arr[i] != largest){
        secondLargest = arr[i];
      }
    }

    return secondLargest;
  }

  /*
  Name: printResult
  Purpose: Print second largest value
    In: int[] arr,
        int answer
    In/Out: N/A
    Out: N/A
  */
  public static void printResult(int answer, int[] arr){

    if (arr.length < 2){
      System.out.println("Less than two values entered.");
    } else {
      System.out.println("The second largest number in the array " + Arrays.toString(arr) + " is: " + answer);
    }
  }

  public static void main(String[] args){

    int[] values = getArray();
    printResult(secondLargest(values),values);
  }
}
